### PR TITLE
Type inference and list initialization

### PR DESCRIPTION
**Pull request by 20joshuaz**
Added support for type inference syntax. Added support for list alloc, alloc + assign and alloc + infer assign syntax.

TODO: Implement multidimensional indexing.

```
a[2][2] is [[1, 2], [3, 4]]. # supported in this PR #
a[0][0]. # not supported; should return 1 #
```

### CODE REVIEW

**Review by Akash-Nayar — APPROVED** [review]
LGTM

File: src/ast.ml [hunk 1]
```diff
@@ -111,18 +115,30 @@ let rec string_of_expr = function
...
+let string_of_array_dimensions dimens = 
+  let array_dimensions_string = String.concat "][" (List.map string_of_expr dimens) in
+  "[" ^ array_dimensions_string ^ "]"
+
+let rec string_of_array array = "[" ^ String.concat ", " (List.map (
+  fun element -> match element with 
+      ExprElem e -> string_of_expr e
+    | ArrayElem a -> string_of_array a
+  ) array
+) ^ "]"
```
Comment (Akash-Nayar) [inline src/ast.ml:133]: good stuff

**Review by Akash-Nayar — COMMENTED** [review]
really good job, just a couple of questions

File: src/ast.ml [hunk 2]
```diff
@@ -32,19 +31,25 @@ type uop =
 type expr = 
     NumberLit of float 
   | BoolLit of bool 
-  | StringLit of string 
   | CharLit of char 
-  | ListLit of expr list 
+  | StringLit of string
```
[the remark this answers is not in the document]
Comment (20joshuaz) [reply src/ast.ml:35]: 😭 😭

File: src/ast.ml [hunk 3]
```diff
@@ -111,18 +115,30 @@ let rec string_of_expr = function
...
+) ^ "]"
 
 let rec string_of_stmt s = 
   let string_of_stmt_raw = function
     Assign (t, id, e) -> string_of_dtype t ^ " " ^ id ^ " is " ^ string_of_expr e ^ ".\n"
-  | Reassign (id, e) -> id ^ " is " ^ string_of_expr e ^ ".\n"
+  | InferAssign (id, e) -> id ^ " is " ^ string_of_expr e ^ ".\n"
+  | Alloc (t, id, dimens) -> string_of_dtype t ^ " " ^ id ^ string_of_array_dimensions dimens ^".\n"
+  | AllocAssign (t, id, dimens, a) -> string_of_dtype t ^ " " ^ id ^ string_of_array_dimensions dimens ^ " is " ^ string_of_array a ^ ".\n"
+  | AllocInferAssign (id, dimens, a) -> id ^ string_of_array_dimensions dimens ^ " is " ^ string_of_array a ^ ".\n"
```
[the remark this answers is not in the document]
Comment (20joshuaz) [reply src/ast.ml:141]: will implement it like
```
array a[5] is [1, 2, 3, 4, 5].
array b[5] is a.
```

[the remark this answers is not in the document]
Comment (20joshuaz) [reply src/ast.ml:141]: Actually I will leave this as a TODO

[the remark this answers is not in the document]
Comment (20joshuaz) [reply src/ast.ml]: Actually I did it.

[the remark this answers is not in the document]
Comment (Akash-Nayar) [reply src/ast.ml]: W mans

**Review by Akash-Nayar — APPROVED** [review]
legend 💯 btw how hard do u think the semant for these changes will be?

**Comment by 20joshuaz** [comment]
> legend 💯 btw how hard do u think the semant for these changes will be?

Should not be too bad!

**Pull request merged**